prow/pubsub/subscriber/subscriber.go

```diff
@@ -141,62 +187,61 @@ func (s *Subscriber) handleMessage(msg messageInterface, subscription string) er
 		s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
 		return err
 	}
+
+	var jh jobHandler
 	switch eType {
 	case periodicProwJobEvent:
-		err := s.handlePeriodicJob(l, msg, subscription)
-		if err != nil {
-			l.WithError(err).Error("failed to create Prow Periodic Job")
-			s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
-		}
-		return err
+		jh = &periodicJobHandler{}
+	case presubmitProwJobEvent:
+		jh = &presubmitJobHandler{}
+	case postsubmitProwJobEvent:
+		jh = &postsubmitJobHandler{}
+	default:
+		l.WithField("type", eType).Error("Unsupported event type")
+		s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
+		return fmt.Errorf("unsupported event type: %s", eType)
+	}
+	if err = s.handleProwJob(l, jh, msg, subscription); err != nil {
+		l.WithError(err).Error("failed to create Prow Periodic Job")
+		s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
 	}
-	err = fmt.Errorf("unsupported event type")
-	l.WithError(err).Error("failed to read message")
-	s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
 	return err
 }
 
-func (s *Subscriber) handlePeriodicJob(l *logrus.Entry, msg messageInterface, subscription string) error {
+func (s *Subscriber) handleProwJob(l *logrus.Entry, jh jobHandler, msg messageInterface, subscription string) error {
 
-	var pe PeriodicProwJobEvent
+	var pe ProwJobEvent
 	var prowJob prowapi.ProwJob
 
+	if err := pe.FromPayload(msg.getPayload()); err != nil {
+		return err
+	}
+
 	reportProwJobFailure := func(pj *prowapi.ProwJob, err error) {
 		pj.Status.State = prowapi.ErrorState
 		pj.Status.Description = err.Error()
-		if s.Reporter.ShouldReport(context.TODO(), l, &prowJob) {
-			if _, _, err := s.Reporter.Report(context.TODO(), l, &prowJob); err != nil {
+		if s.Reporter.ShouldReport(context.TODO(), l, pj) {
+			if _, _, err := s.Reporter.Report(context.TODO(), l, pj); err != nil {
 				l.Warningf("failed to report status. %v", err)
 			}
 		}
 	}
 
-	if err := pe.FromPayload(msg.getPayload()); err != nil {
-		return err
-	}
-	var periodicJob *config.Periodic
-	for _, job := range s.ConfigAgent.Config().AllPeriodics() {
-		if job.Name == pe.Name {
-			periodicJob = &job
-			break
-		}
-	}
-	if periodicJob == nil {
-		err := fmt.Errorf("failed to find associated periodic job %q", pe.Name)
+	prowJobSpec, labels, err := jh.getProwJobSpec(s.ConfigAgent.Config(), pe)
+	if err != nil || prowJobSpec == nil {
```

Comment:
It only makes sense to check if `prowJobSpec` is nil if it is possible for both `err` and `prowJobSpec` to be nil. That isn't possible currently and probably shouldn't be allowed. If it is expected to be allowed then we should update this to populate `err` in that case, otherwise we can return nil in an error case.

Reply:
good catch